File: meta_analysis_geneexpr.py
#!/usr/bin/python
import argparse
import sys
import subprocess
import os

def lsf_file(job_name, cmd, memory=36000, thread=1):
    """
    Creates .lsf files
    """

    outp = open(job_name+".lsf",'w')
    outp.write("#!/bin/bash\n")
    outp.write("#BSUB -L /bin/bash\n")
    outp.write("#BSUB -J "+job_name+"\n")
    outp.write("#BSUB -q normal\n")
    outp.write("#BSUB -o "+job_name+"_%J.out\n")
    outp.write("#BSUB -e "+job_name+"_%J.screen\n")
    outp.write("#BSUB -M "+str(memory)+"\n")
    outp.write("#BSUB -n "+str(thread)+"\n")
    outp.write(cmd)
    outp.write("\n")
    outp.close()



def main(samp_info, DE_dir, script_dir, study, tissue, disease, treatment, out, method):
    """
    Run meta-analysis using meta_analysis_microarry.R
    """

    # check exists of script directory
    if script_dir == "./":
        script_dir = os.getcwd()
    if script_dir[-1] != "/":
        script_dir = script_dir+"/"

    out_dir=os.path.dirname(out)
    if not os.path.exists(out_dir):
        os.makedirs(out_dir)
        print "Create output directory" +out_dir

    if not os.path.isfile(script_dir+"meta_analysis_geneexpr.R"):
        print "R script file meta_analysis_geneexpr.R does not exist in "+script_dir
        sys.exit()

    if not os.path.isfile(script_dir+"integration_utility.R"):
        print "R utility script file integration_utility.R does not exist in "+script_dir
        sys.exit()

    if method=="rankprodt":
        if not os.path.isfile(script_dir+"meta_analysis_RankProd.R"):
            print "R script file meta_analysis_RankProd.R for rank product permutation does not exist in "+script_dir
            sys.exit()

    if (disease is not None) and (treatment is not None):
        print "Both disease endotype and treatment are specified. The runtime will be doubled."
        print "You can just specify disease endotype or treatment to reduce the runtime"

    if study is None:
        study="NA"
    else:
        tissue="NA"
        disease="NA"
        treatment="NA"

    if disease is None:
        disease="NA"

    if treatment is None:
        treatment="NA"

    cmd="Rscript --vanilla "+script_dir+"meta_analysis_geneexpr.R \""+samp_info+"\" \""+DE_dir+"\" \""+script_dir+"\" \""+study+"\" \""+tissue+"\" \""+disease+"\" \""+treatment+"\" \""+out+"\" \""+method+"\""
    print "Run R commad..."
    print cmd
    os.system(cmd)

    # Create 1000 .lsf files for 1,000,000 permutations
    if method=="rankprodt":
        if not os.path.exists(out+".ranklist.RDS"):
            print "Input rank list RDS file "+out+".ranklist.RDS"+" does not exist."
            sys.exit()

        # create a directory to put subpermutation files
        new_dir=out+"_rankperm"
        if not os.path.exists(new_dir):
            os.makedirs(new_dir)
            print "Create new directory "+new_dir+" to save sub-permutation results"
        else:
            print "Directory "+new_dir+" already exists. Please check!"
            sys.exit()

        print "Create 20 .lsf files..."
        # create 20 LSF files for 1000 sub-permutation
        seeds=range(1,1000000,50000) # 20 seeds: 1, 5001, 10,001,...,950,001
        i=0
        # number of cores used
        ncore=10
        for seed in seeds:
            i=i+1
            # R script arguments: V1: script dir, V2: input RDS, V3: nperm in sub-permutation, V4: random seed, V5: number of cores (use 10), V6: output prefix, usually named by seeds, V7: result directory
            out_prefix=str(seed)
            cmd="time Rscript --vanilla "+script_dir+"meta_analysis_RankProd.R "+script_dir+" "+out+".ranklist.RDS 50000 "+str(seed)+" "+str(ncore)+" "+" "+out_prefix+" "+new_dir # use seed as output prefix
            lsf_file("rank_"+str(seed), cmd, memory=24000, thread=ncore)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Meta-analysis of transcriptomic data")
    parser.add_argument("--samp_info", type=str, help="sample info sheet .RDS for all studies")
    parser.add_argument("--DE_dir", type=str, default="./", help="directory with .RDS files of DE results")

    parser.add_argument("--script_dir", type=str, default="./", help="directory to put R script files meta_analysis_geneexpr.R and integration_utility.R for meta-analysis")
    parser.add_argument("--study", type=str, help="study based on Unique_ID from Sample Info Sheet. If specified, ignore other filtrations")
    parser.add_argument("--tissue",type=str, help="tissue: entire or ASM BAL BE CD4 CD8 chALL LCL LEC Lung MACRO MCF10A-Myc NE PBMC SAE U2OS WBC. Use quote for multiple tissues")
    parser.add_argument("--disease", type=str, help="disease: entire or allergic_asthma, asthma, asthma_and_rhinitis, fatal_asthma, non_allergic_asthma, non_severe_asthma, severe_asthma. Use quote for multiple disease endotype(s)")
    parser.add_argument("--treatment", type=str, help="treatment: GC BA_GC vitD. Use quote for treatment(s)")
    parser.add_argument("--out", type=str, default="out", help="output file name prefix including directory path (default=out)")
    parser.add_argument("--method", type=str, default="metaranef", help="Methods used for integration. Options: metaranef, fisherp, rankprodt (default=metaranef)")

    args = parser.parse_args()

    if len(sys.argv)==1:
        parser.print_help()
        sys.exit(1)

    if args.samp_info is None:
        print "--samp_info option is required"
        parser.print_help()
        sys.exit(1)
        
    if args.DE_dir is None:
        print "--DE_dir option is required"
        parser.print_help()
        sys.exit(1)

    if args.study is None:
        if args.tissue is None:
            print "Please specify tissue(s)."
            sys.exit(1)

        if (args.disease is None) and (args.treatment is None):
            print "Please specify disease endotype(s) or treatment(s)"
            sys.exit(1)

    else:
        print "Selection only based on unique ID. Ignore other specified filtrations, e.g. tissue, disease, treatment"

    if args.method not in ["metaranef", "fisherp", "rankprodt"]:
        print "Please check method options."
        sys.exit(1)

    main(args.samp_info, args.DE_dir, args.script_dir, args.study, args.tissue, args.disease, args.treatment, args.out, args.method)

File: meta_analysis_geneexpr.R
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

###
# Usage
###

# Select datasets and run integration analysis. Variables are imported from meta_analysis_geneexpr.py

samp_info=args[1]
DE_dir=args[2]
script_dir=args[3]
study=args[4]
tissue=args[5]
disease=args[6]
treatment=args[7]
out=args[8]
method=args[9]

cat("samp_info =", samp_info,"\n")
cat("DE_dir =", DE_dir,"\n")
cat("script_dir =", script_dir,"\n")
cat("study =", study,"\n")
cat("tissue =", tissue,"\n")
cat("disease =", disease,"\n")
cat("treatment =", treatment,"\n")
cat("out =", out,"\n")
cat("method =", method,"\n")

# utility file
#script_dir = "/home/mengykan/Projects/asthmagene/scripts/meta-analysis"
utility_fn=paste0(script_dir,"integration_utility.R")

###
# Change Here
###

# gene expression data path and files
geneexpr_dir=DE_dir # "/project/bhimeslab/databases/AsthmaApp/databases/Microarray_data_infosheet_R.RDS"
geneexpr_datinfo_fn=samp_info # "/project/bhimeslab/databases/AsthmaApp/databases/microarray_results"

if (!(dir.exists(geneexpr_dir))) stop(paste0("The gene expression data directory "),geneexpr_dir," does not exist")
if (!(file.exists(geneexpr_datinfo_fn))) stop(paste0("The gene expression data information file "),geneexpr_datinfo_fn," does not exist")
datinfo=readRDS(geneexpr_datinfo_fn)

# output date varialbe for output filename creation
# date=Sys.Date()
# date=gsub("-","",date)
date="" # no longer use date in filenames

# load utility resource
if (!(file.exists(utility_fn))) stop(paste0("The utility file "),utility_fn," does not exist")
source(utility_fn)

# Convert arguments into variables
if (study=="NA") {
  tissue=strsplit(tissue, "\\s")[[1]]

  if (disease!="NA") {
    disease=strsplit(disease, "\\s")[[1]]
  }
  if (treatment!="NA") {
    treatment=strsplit(treatment, "\\s")[[1]]
  }
} else {
  study=strsplit(study, "\\s")[[1]]
}

# create output file name prefix and related directory
geneexpr_out_prefix=out

rankmethod_func <- function(datinfo, genes) {
  # create rank list R object for rank-based integration
  rankdat_list=lapply(readinlist_func(datinfo),function(x){x[x$Gene%in%genes,c("Gene","rank")]})
  # perform permutation
  # res=rankprodt_perm(dat.list=rankdat_list,nperm=1000,rand=1)
  # As we need 1000,000 permutations, create 20 jobs each with 50,000 permutations and run them separately
  ranklist_fn=paste0(geneexpr_out_prefix,".ranklist.RDS")
  if (file.exists(ranklist_fn)) cat("The ranklist file already exists and is overwritten now.\n")
  saveRDS(rankdat_list,ranklist_fn)
  cat("Obtain ranklist object.\n")
  # compute real rank product
  dat <- Reduce(function(x,y){merge(x,y,by="Gene",all=T)},rankdat_list)
  # compute real rank product
  rankprod=apply(dat[,!names(dat)%in%"Gene"],1,rankprodt)
  res=data.frame(Gene=dat$Gene,rankprod=rankprod)
  write.table(res[order(res$rankprod),],paste0(geneexpr_out_prefix,".rankprodt.txt"),sep="\t",row.names=F,col.names=T,quote=F)
}

fisherpmethod_func <- function(datinfo, genes) {
  # convert p-values based on the effect direction
  cat("Convert p-values based on the effect direction\n")
  res_list=pvalconvt(datinfo=datinfo, gene=genes, col_var="P.Value", mode="gene", ncore=10) # convert p-values based on the effect direction, get genes
  # run Fisher's combination for one-tail p-values and save data
  cat("Compute fisher's p-values\n")
  score=genescore_singletype(res_list,col_var="stats",func="fisherp")
  score$P.Value <- score$Score
  score$qval <- p.adjust(score$Score,method="BH")
  score$rank <- rank(score$Score)
  score$Score <- NULL
  score <- score[order(score$rank),]
  write.table(score,paste0(geneexpr_out_prefix,".fisherp.txt"),sep="\t",row.names=F,col.names=T,quote=F)
}


metaranefmethod_func <- function(datinfo, genes) {
  # obtain list by gene
  if ("Population"%in%names(datinfo)) {data_dir=gwas_dir} else {data_dir=geneexpr_dir}
  study_list=list()
  for (i in datinfo$Unique_ID) {study_list[[i]]=readRDS(paste0(data_dir, "/", i, ".RDS"))}
  cat("Perform meta-analysis using random-effects model\n")
  res_list=lapply(genes,function(gene){
    d=sapply(study_list,function(dat){if(gene%in%dat$Gene){dat[which(dat$Gene%in%gene),"logFC"]}else{NA}})
    se=sapply(study_list,function(dat){if(gene%in%dat$Gene){dat[which(dat$Gene%in%gene),"SD"]}else{NA}})
    metaranef(d=d,se=se,method="random")
  })
  score=do.call(rbind,res_list)
  score=data.frame(Gene=genes,score)
  score$qval <- p.adjust(score$P.Value,method="BH")
  score$rank <- rank(score$P.Value)
  score <- score[order(score$rank),]
  write.table(score,paste0(geneexpr_out_prefix,".metaranef.txt"),sep="\t",row.names=F,col.names=T,quote=F)
}


meta_func <- function(geneexpr_datinfo, method){

  # select genes shared within at least two studies
  cat("Select genes shared within at least two studies\n")
  genes=sharegene_func(dat.list=readinlist_func(geneexpr_datinfo),minnum=2)
  cat("Obtain",length(genes),"genes\n")

  # genes=c("CEBPD","TSC22D3") # two genes for testing

  ###
  # Rand product
  ###
  if (method=="rankprodt") {
    rankmethod_func(datinfo=geneexpr_datinfo, genes=genes)
  }

  ###
  # Fisherp
  ###

  if (method=="fisherp") {
    fisherpmethod_func(datinfo=geneexpr_datinfo, gene=genes)
  }

  ###
  # Random-effects model
  ###
  if (method=="metaranef") { 
    metaranefmethod_func(datinfo=geneexpr_datinfo, genes=genes)
  }
}

# Study
if (!"NA"%in%study) {
  # select gene expression study based on Unique_ID
  geneexpr_datinfo=geneexpr_study_select(study)
  print(geneexpr_datinfo)
  if (nrow(geneexpr_datinfo)<2) {cat("Less than 2 gene expression studies selected. Cannot perform integration.\n")} else {meta_func(geneexpr_datinfo=geneexpr_datinfo, method=method)}
}


# Asthma
if (!"NA"%in%disease) {
  # select disease gene expression study
  geneexpr_datinfo=geneexpr_disease_select(tissue,disease)
  print(geneexpr_datinfo)
  if (nrow(geneexpr_datinfo)<2) {cat("Less than 2 disease gene expression studies. Cannot perform integration.\n")} else {meta_func(geneexpr_datinfo=geneexpr_datinfo, method=method)}
}

# Treatment
if (!"NA"%in%treatment) {
  # select treatment gene expression study
  geneexpr_datinfo=geneexpr_GC_select(tissue,treatment)
  print(geneexpr_datinfo)
  if (nrow(geneexpr_datinfo)<2) {cat("Less than 2 treatment gene expression studies. Cannot perform integration.\n")} else {meta_func(geneexpr_datinfo=geneexpr_datinfo, method=method)}
}
